Adding fields selector

sdminonne opened the pull request:
@bgrant0607 @nikhiljindal @smarterclayton 

Goal of this PR is introduce fields type. This is done simply copying old simple implementation of labels.
We briefly discussed about this here: https://github.com/GoogleCloudPlatform/kubernetes/issues/341#issuecomment-75712472

Most of the changes are quite 'natural'. 

Please double check: pkg/client/request.go: ParseSelectorParam: not sure if this is good idiomatic golang.
One of the missing part is the cleanup of the old implementation of labels (based on `hasTerm/andTerm`). In the next days I'll submit another PR to clean this.

sdminonne committed:
Adding fields selector

bgrant0607 commented:
Overall approach is what I expected. Adds another top-level directory under pkg, but that's a general problem, and shouldn't block this PR.

Reassigning for detailed review.

nikhiljindal commented:
cc @lavalamp

nikhiljindal commented on pkg/fields/selector_test.go:
```diff
@@ -0,0 +1,212 @@
+package fields
+
+import (
+	//	"reflect"
```
Remove unused imports

nikhiljindal commented:
Right now, there is a lot of code duplicated in labels and fields. As per your description it seems that we will be deleting all of it from labels? Its fine if thats the case, else we need a way to share the common code.

nikhiljindal commented:
LGTM with a minor nit to delete unused imports.

lavalamp commented:
What is pkg/fields for?

sdminonne commented:
@lavalamp working on https://github.com/GoogleCloudPlatform/kubernetes/issues/341 I realized that `fields` and `labels` have fundamentally the same go type while they have different syntax  and different goals. See 
 https://github.com/GoogleCloudPlatform/kubernetes/issues/341#issuecomment-75516637, the outcome is this PR.

sdminonne commented:
@nikhiljindal as I was mentioning I'm going to remove duplicated code in the next hours, and sharing code if possible. Thanks

sdminonne commented:
@nikhiljindal @bgrant0607 @lavalamp. Just updated.

nikhiljindal commented:
Thanks, LGTM!
You need to make travis happy :)

sdminonne commented:
@nikhiljindal sorry :)

lavalamp commented:
> @lavalamp working on #341 I realized that fields and labels have fundamentally the same go type while they have different syntax and different goals. See 
#341 (comment), the outcome is this PR.

@sdminonne The distinction you make in that comment is between selectors and labels, not between fields and labels. AFAIK, selectors should work on both fields and labels.

smarterclayton commented:
The syntax of fields and labels would differ - does that also imply field selectors and label selectors are not equivalent?


> On Mar 10, 2015, at 2:51 PM, Daniel Smith <notifications@github.com> wrote:
> 
> @lavalamp working on #341 I realized that fields and labels have fundamentally the same go type while they have different syntax and different goals. See 
> #341 (comment), the outcome is this PR.
> 
> @sdminonne The distinction you make in that comment is between selectors and labels, not between fields and labels. AFAIK, selectors should work on both fields and labels.
> 
> —
> Reply to this email directly or view it on GitHub.
>

bgrant0607 commented:
I suggested this approach in order to facilitate progress on label selectors. 

At minimum, the validation rules and parsing would be different for fields -- at this point, I think field keys are more restrictive than labels (e.g., no `/`) but field values are less restrictive (i.e., they have no consistent restrictions), and may include non-string values. 

I could imaging re-unifying the implementations at some point, but I don't think it's a priority at the moment, especially since we haven't decided how to handle field selection more generally.

sdminonne commented:
@lavalamp sorry for the misunderstanding: in my comment was related to the different selector syntax which (IMHO) should produce different types. 
Avoiding bugs like `func f(labels, fields labels.Selector)` called with swapped input param `f(label, field)` VS `f(field, label)`...
Finally I agree with  @smarterclayton 
PS: rebased and squashed since I'm not lucky with shippable

bgrant0607 commented:
Shippable passed. Happy to re-evaluate merging in subsequent PRs.
